### ISSUE TITLE
Code using SDL 2.0 will not compile to WASM

### ISSUE THREAD

**Issue by pbferre** [emscripten-core/emscripten#5097]
I had an issue where I was having trouble displaying image files created in the app on the browser page. I solved it by switching from SDL 1.2 to 2.0. However, the new code will not compile to WASM. It compiles to asm.js, although I can only use -O1 optimization. However, if I add the '-s WASM=1' switch, the asm2wasm.exe crashes with this message
```
Assertion failed: mappedGlobals.find(name) != mappedGlobals.end() ? true : (std::cerr << name.str << '\n', false), file C:\Users\Patrick Ferre\.emscripten_ports\binaryen\binaryen-version_30\src\asm2wasm.h, line 1575
```

Once I close the program, I get this traceback
```
Traceback (most recent call last):
  File "C:\Program Files\Emscripten\emscripten\incoming\\emcc", line 13, in <module>
    emcc.run()
  File "C:\Program Files\Emscripten\emscripten\incoming\emcc.py", line 2170, in run
    subprocess.check_call(cmd)
  File "C:\Program Files\Emscripten\python\2.7.5.3_64bit\lib\subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '[u'C:\\Users\\Patrick Ferre\\.emscripten_ports\\binaryen\\binaryen-version_30\\bin\\asm2wasm', 'dither2.temp.asm.js', '--total-memory=16777216', '--emit-jsified-potential-traps', '-O1', '--mem-init=dither2.html.mem', '--mem-base=1024', '--wasm-only', '-o', 'dither2.wasm']' returned non-zero exit status -1073740791
```

Is there an issue with SDL 2.0, and the WASM compiler, or could this possibly be something in my code? It compiles the individual units correctly, into object files. I only get this error in the last step, combining the object files into the final html.

**Comment by kripken** [comment]
You may be hitting an unknown bug in asm2wasm. Is that on very latest incoming, first thing I'd check? If so, can you provide the input to asm2wasm (the .asm.js file, emitted when EMCC_DEBUG is on).

**Comment by pbferre** [comment]
There is a  .temp.asm.js file, but it is 39 MB in size. I cannot attach it.

**Comment by kripken** [comment]
Perhaps zipped it is small enough? Or you can upload it to another site, or email it?

**Comment by pbferre** [comment]
I have been experimenting with different versions of the SDK, attempting to diagnose this, and I have encountered some issues which may be of use to you going forward. I installed 64 bit versions of SDKs 1.37.1, 1.37.8, and 1.37.9. 

1.37.9 will not compile to WASM at all. I get this, when it is trying to build wasm-libc.bc
```
WARNING:root:Accessing the Emscripten cache at "C:\Users\Patrick Ferre\.emscripten_cache\asmjs" is taking a long time, another process should be writing to it. If there are none and you suspect this process has deadlocked, try deleting the lock file "C:\Users\Patrick Ferre\.emscripten_cache.lock" and try again. If this occurs deterministically, consider filing a bug.
```
1.37.1 and 1.37.9 will compile to WASM, but not if I add support for .png files. I changed the line in settings.js to read **var SDL2_IMAGE_FORMATS = ["png"];**. That is when I get the assertion failed error described above, but again, only of that line in settings.js has been changed. If I do the same thing in 1.37.1, it still will not compile to WASM if that line has been altered, but the error is different.
```
parse exception: bad get_global name (at 2123813:16)]Fatal: error in parsing input
Traceback (most recent call last):
  File "C:\Program Files\Emscripten\emscripten\1.37.1\\emcc", line 13, in <module>
    emcc.run()
  File "C:\Program Files\Emscripten\emscripten\1.37.1\emcc.py", line 2136, in run
    subprocess.check_call(cmd)
  File "C:\Program Files\Emscripten\python\2.7.5.3_64bit\lib\subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '[u'C:/Program Files/Emscripten/clang/e1.37.1_64bit/binaryen\\bin\\wasm-as', 'dither2.wast', '-o', 'dither2.wasm']' returned non-zero exit status 1
```

I cannot add support for png files to the command line at all. In every version of the SDK, if I add the switch **-s SDL2_IMAGE_FORMATS='["png"]'** to the emcc command, I get this failure.
```
ERROR:root:a problem occurred when using an emscripten-ports library. try to run    emcc --clear-ports    and then run this command again
Traceback (most recent call last):
  File "C:\Program Files\Emscripten\emscripten\1.37.9\\emcc", line 13, in <module>
    emcc.run()
  File "C:\Program Files\Emscripten\emscripten\1.37.9\emcc.py", line 1540, in run
    extra_files_to_link = system_libs.get_ports(shared.Settings)
  File "C:\Program Files\Emscripten\emscripten\1.37.9\tools\system_libs.py", line 620, in get_ports
    ret += filter(lambda f: not f.endswith('.txt'), port.get(Ports, settings, shared))
  File "C:\Program Files\Emscripten\emscripten\1.37.9\tools\ports\sdl_image.py", line 34, in get
    settings.SDL2_IMAGE_FORMATS.sort()
AttributeError: 'str' object has no attribute 'sort'
```

I ran emcc --clear-ports and tried again. No difference.

So to sum up, there is definitely an issue with compiling to WASM if using SDL2, and attempting to support png files.

**Comment by pbferre** [comment]
Sorry, the first sentence in the second paragraph should read "1.37.1 and **1.37.8** will compile to WASM, but not if I add support for .png files."

**Comment by kripken** [comment]
Yeah, the bug you are hitting must be very specific to using wasm + sdl2 + png files.

**Comment by pbferre** [comment]
Yeah. Not to keep raining on you, but I discovered another really obscure one you might want to know about. The first thing my app does is prompt for a url of the image to download. Running in a browser, that means a window pops to prompt you. Now with optimization level of -O1 or lower, this works fine across all browsers. If I use level -O2 or higher, it works on Firefox, but on Chrome and Edge, I never get the prompt for input, just sits there spinning its heels. This is irregardless of whether I use C++ cin statements or C fgets. Just something you may want to eventually look at.

**Comment by kripken** [comment]
Reproducible testcases are necessary for us to look at all those issues, both the main compile error and the other stuff you mention.

Btw, in addition to the other options I mentioned before for submitting a large testcase, you can create a new github repo with it (+steps to reproduce etc., in the repo).

**Comment by pbferre** [comment]
I'll look into it.

**Comment by atrosinenko** [comment]
@pbferre As far as I know, `libpng` uses `setjmp` / `longjmp` for error handling. In case you are using Asyncify, then it may be connected with #5383 (in fact, the symbol that caused the crash was printed _just before_ the `Assertion failed` line).

**Comment by kripken** [comment]
Note that asyncify is not recommended, the emterpreter option is maintained and can provide most of the same use cases (except for coroutines).

**Comment by atrosinenko** [comment]
Now coroutines should work in Emterpreter too.
